nixos/netbird: harden and extend options

nazarewk opened the pull request:
## Description of changes

I have recently extensively tested and fixed all features of Netbird in  [my own implementation](https://github.com/nazarewk-iac/nix-configs/blob/fb8bac011db7075b9c4f473ae4490a2ac727b29c/modules/networking/netbird/default.nix) of multi-instance Netbird installations.

While doing so I discovered another multi-instance implementation got merged into nixpkgs https://github.com/NixOS/nixpkgs/pull/246055 which is slightly different, but still a solid base to upstream the rest of my changes:
- running as ~~DynamicUser~~ it's own user with minimal set of permissions
	- it was there before, but was lacking some of capabilities,
- made some configurations situational
- add more *unmanaged* interface configurations
- quality of life improvements:
	- configure log level for each interface
	- optionally turn off starting during boot
	- `openFirewall` by default
	- add shortcuts/wrappers for each created instance

I think it's a pretty good time to upstream, because I will be extensively using it at work: just launched my first Colmena-managed NixOS into GCE.

There are plans to support multi-account connections on the same daemon in Q2/2024 (see [the slack message](https://netbirdio.slack.com/archives/C02KHAE8VLZ/p1706603733407369?thread_ts=1706566734.905759&cid=C02KHAE8VLZ)), but it's not known what shape it will take at all.

I decided to implement following significant changes:
- instances must specify a port they will be listening on as it doesn't make much sense to give an immediately conflicting default,
- aliased `tunnels` to `clients`, because a word `tunnel` does not exist in Netbird's nomenclature ([unlike some other VPNs](https://search.nixos.org/options?channel=23.11&from=0&size=50&sort=relevance&type=packages&query=tunnels)) and is pretty misleading. Also `clients.*` play nicely with my plan to implement a `server` in near future.
- skipped destructuring expressions  (eg: `{name, ...}: name` -> `client: client.name`) because they make the code very hard to follow and update with increased number of options,

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

misuzu commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,383 @@
   };
 
   config = mkMerge [
-    (mkIf cfg.enable {
```
I think we should keep this. Once multi-account connections are implemented in netbird itself, the update should be a no-op for the majority of users. There's also a section for NixOS in their documentation, we should keep supporting it: https://docs.netbird.io/how-to/getting-started#linux

nazarewk replied:
nice find on that `meta.doc`, will also help me documenting the other PR I have opened :)

misuzu commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,383 @@
     thubrecht
+    nazarewk
   ];
   meta.doc = ./netbird.md;
```
All changes to the API must be reflected in the documentation

misuzu commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,383 @@
+                      ${exports}
+                      ${lib.getExe' cfg.package tool} "$@"
+                    '';
+                    tools = [ "netbird" "netbird-mgmt" "netbird-signal" ];
```
Only `netbird` is needed for the client, everything else is for the server

misuzu commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,383 @@
+                description = "log level of this client";
+              };
+
+              firewall.tcp.ports = mkOption {
```
Why is the `firewall` option needed? The description is not informative

nazarewk replied:
The idea was to allow opening up firewall for services without having to figure out the interface naming conventions.

Having `config.interface` available makes it "good enough" to use with standard firewall module, so I could remove it.

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.md:
````diff
@@ -2,55 +2,98 @@
 
 ```nix
-services.netbird.enable = true;
+services.netbird.clients.default.port = 51820;
````
The simplest config is still `services.netbird.enable = true;`

Tom-Hubrecht commented on nixos/tests/netbird.nix:
```diff
@@ -12,6 +12,7 @@ import ./make-test-python.nix ({ pkgs, lib, ... }:
     };
   };
 
+  # TODO: confirm DynamicUser is working if/after netbird server is implemented
```
There is no netbird server in this PR

nazarewk replied:
yeah, but it is very much relevant for the test, there is no confirmation the client works at all until we do have a server running.

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,418 @@
+                internal = true;
+                default =
+                  let
+                    exports = pipe config.environment [
+                      (mapAttrsToList (name: default:
+                        # Couldn't default with `''${${name}:${escapeShellArg default}}`
+                        # because `escapeShellArg` wraps values in single quotes,
+                        # which in turn works differenty (or rather doesn't) than double-quotes
+                        # in this shell parameter expansion
+                        ''test "''${${name}+X}" == X || export ${toShellVar name default}''))
+                      (concatStringsSep "\n")
+                    ];
+                  in
+                  pkgs.writeShellScriptBin "netbird${config.suffix}" ''
+                    ${exports}
+                    ${lib.getExe' cfg.package "netbird"} "$@"
+                  '';
```
This is very weird, nixpkgs has a machinery already to wrap programs c.f. https://nixos.org/manual/nixpkgs/stable/#fun-wrapProgram

nazarewk replied:
I'm pretty sure it would be significantly longer with `wrapProgram`, something like:
```
mkDerivation {
	pname = "xxx";
    installPhase = 
	    let
	  		exports = pipe config.environment [
	          (mapAttrsToList (name: default:
	            "--set-default '${name}' '${default}'"))
	          (concatStringsSep " ")
	        ];
		in
		"wrapProgram ${lib.getExe' cfg.package "netbird"} $out/bin/netbird${config.suffix} ${exports}";
}
```

actually makes more sense after prototyping it and seeing with own eyes

Tom-Hubrecht replied:
You can do something like
```nix
    stdenv.mkDerivation {
      intallPhase = ''
        wrapProgram ${lib.getExe cfg.package} $out/bin/netbird${config.suffix} ${
          concatMapStringsSep " " ({ name, value }: "--set-default '${name}' '${default}'") (
            attrsToList config.environment
          )
        }
      '';
```
The issue I had was not because of the size of the code, but reinventing the wheel ^^

nazarewk replied:
not so short anymore:
```nix
let
  makeWrapperArgs = flatten (mapAttrsToList
    (key: value: [ "--set-default" key value ])
    config.environment
  );
in
pkgs.stdenv.mkDerivation {
  name = "${cfg.package.name}-wrapper-${name}";
  meta.mainProgram = config.name;
  nativeBuildInputs = with pkgs; [ makeWrapper ];
  phases = [ "installPhase" ];
  installPhase = ''
    mkdir -p "$out/bin"
    makeWrapper ${lib.getExe cfg.package} "$out/bin/${config.name}" \
      ${escapeShellArgs makeWrapperArgs}
  '';
}
```

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,418 @@
+
+        A single daemon is fully operational.
+
+        Multiple daemons will interfere with each other's DNS resolution, but otherwise should remain operational.
```
When the networks have different domains, there is no interference, the only issue is when then share the same domain

nazarewk replied:
I don't think the domain is configurable at all on the Netbird server side?

Tom-Hubrecht replied:
It is when you self-host your server

nazarewk replied:
FYI:
> Hello[@nazarewk, we plan a custom DNS zone support coming next quarter. That would allow you to create custom domains for applications and peers

https://netbirdio.slack.com/archives/C02KHAE8VLZ/p1707820381270479?thread_ts=1707819591.188209&cid=C02KHAE8VLZ

CertainLach replied:
Custom DNS zone support is already implemented for self-hosted,
```sh
        ${pkgs.netbird}/bin/netbird-mgmt management \
        --dns-domain dns.domain
```
Will place every machine under dns.domain subdomain: `machine1.dns.domain`, `machine2.dns.domain`

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,418 @@
-              inherit environment;
+            # see https://github.com/netbirdio/netbird/issues/1079#issuecomment-1823180247
+            CacheDirectory = client.name;
+            ConfigurationDirectory = client.name;
```
Netbird should not use anything under `/etc` so you should no set `ConfigurationDirectory`

nazarewk replied:
yeah, dropped it as too irrelevant to current state of affairs

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,418 @@
+            Restart = "always";
 
-              inherit environment;
+            # see https://github.com/netbirdio/netbird/issues/1079#issuecomment-1823180247
```
I fail to see how this issue is relevant to our case, you don't set `LogDirectory` as it is unused when using console and thus will not happen here

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,418 @@
+
+                echo "Checking hardened=$hardened specific permissions..." >&2
+
+                ensure_owner() {
+                  local dir="$1" repr="$2"
+                  echo "Checking ownership of $repr" >&2
+
+                  while IFS= read -r file; do
+                    if [ "$(stat -c "%u:%g" "$file")" == "$owner" ] ; then
+                      echo "- OK:     $file already owned by $owner" >&2
+                    else
+                      echo "- CHANGE: $file changing owner to $owner" >&2
+                      chown "$owner" "$file"
+                    fi
+                  done < <(find "$dir")
+
+                }
+
+                for dir in "''${dirs[@]}" ; do
+                  [[ -e "$dir" ]] || continue
+
+                  repr="$dir"
+                  realdir="$(realpath "$dir" || echo -n "$dir")"
+
+                  if [ "$dir" != "$realdir" ] ; then
+                    ensure_owner "$dir" "$repr"
+                    repr="$dir (-> $realdir)"
+                    dir="$realdir"
+                  fi
+                  ensure_owner "$dir" "$repr"
+                done
+              '';
+            });
+            Type = "oneshot";
+            User = "root";
+            RemainAfterExit = "yes";
+            Restart = "no";
+          };
+        }
+      );
```
I don't really get what you are trying to do with this whole bash blob, systemd should set the permissions correctly at startup

nazarewk replied:
it does so only for the first start of the service, not when toggling between hardened/not hardened

nazarewk replied:
actually seems like it does the right thing as long as the `Directory` itself is owned by the "wrong" user

Tom-Hubrecht replied:
So as for now everything is owned by root, systemd will do it for us automatically

nazarewk replied:
yeah, I'm confirming it on my branch, but got some stuff messed up due too many open PRs (merged into that branch)

nazarewk replied:
It seems to work for all directories except `ConfigurationDirectory`

nazarewk commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,383 @@
+                '';
+              };
+
+              suffix = mkOption {
+                type = str;
+                default = "-${name}";
+                description = ''
+                  Suffix to use for this further naming of the service pieces.
+                '';
+              };
```
It is also used for the wrapper, but yeah might make more sense to not drop `-wt0` from the default config. The raw "unwrapped" package should work with it anyway.

nazarewk replied:
it made more sense when there were server wrappers

nazarewk commented:
Identified following things:
- Netbird actually puts config in `/etc/netbird/config.json` by default
	- will add migration script to cover it
- forgot to open up firewall for direct p2p communacation over LAN

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,336 @@
+
+              openFirewall = mkOption {
+                type = bool;
+                default = true;
```
I would set the default value to `false`, the opening of a port in the firewall should be a conscious decision

nazarewk replied:
I'd argue it doesn't make much sense setting it up to connect to LAN peers over the internet by default while the port is securely authenticated anyway.

There is some precedence for secure connection services:
- https://github.com/NixOS/nixpkgs/blob/f9d39fb9aff0efee4a3d5f4a6d7c17701d38a1d8/nixos/modules/services/networking/wgautomesh.nix#L51
- https://github.com/NixOS/nixpkgs/blob/f9d39fb9aff0efee4a3d5f4a6d7c17701d38a1d8/nixos/modules/services/networking/ssh/sshd.nix#L220
- https://github.com/NixOS/nixpkgs/blob/f9d39fb9aff0efee4a3d5f4a6d7c17701d38a1d8/nixos/modules/programs/mosh.nix#L13

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,336 @@
   meta.doc = ./netbird.md;
 
+  imports = [
+    (mkAliasOptionModule [ "services" "netbird" "tunnels" ] [ "services" "netbird" "clients" ])
```
```suggestion
    (mkRemovedOptionModule [ "services" "netbird" "tunnels" ] "The option `services.netbird.tunnels` has been renamed to `services.netbird.clients`")
```
I don't think that many people have already started to use the `tunnels` option, not much disturbance would happen with the removal

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.md:
````diff
@@ -2,55 +2,113 @@
 ```
 
-The address will by default be `unix:///var/run/netbird-<name>`.
+## Configuration interpolation {#module-services-netbird-interpolation}
````
The term interpolation seems weird to me in this context, maybe use `interface-specific` ?

nazarewk replied:
yeah, I'm not sure what to call it either, how about *Configuration overview*?

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.md:
```diff
@@ -2,55 +2,113 @@
+
+Each Netbird client service:
+
+- by default runs in `hardened = true` configuration,
+- by default starts with the system `autoStart = true`,
+- can be additionally configured with `environment`,
```
```suggestion
- runs by default in a hardened mode,
- starts by default alongside the sytem,
- can be additionally configured with environment variables,
```

Tom-Hubrecht commented on nixos/modules/services/networking/netbird.md:
````diff
@@ -2,55 +2,113 @@
 ```
 
-This will set the socket to interact with the netbird service to `/var/run/toto.sock`.
+Above configuration will:
+
+1. start the Netbird daemon in `netbird-wt1.service`
+2. manage `netbird-wt1` network/wireguard interface
+3. listen for incoming remote connections on the port `51830`
+4. use the `/var/lib/netbird-wt1/config.json` configuration file
+5. will be locally manageable on `unix:///var/run/netbird-wt1/sock` socket
+6. will have a `netbird-wt1` shortcut as alternative to `netbird --daemon-addr unix:///var/run/netbird-wt1/sock`
````
```suggestion
The daemon started via this configuration will:

1. run as the `netbird-wt1.service`
2. manage the `netbird-wt1` network interface
3. listen for incoming remote connections on the port `51830`
4. use the `/var/lib/netbird-wt1/config.json` configuration file
5. be locally manageable through the `unix:///var/run/netbird-wt1/sock` socket

A `netbird-wt1` shortcut will be made available to interact with the daemon using the correct settings.
```

misuzu commented:
Checklist for manual testing:
* DNS management should work out of the box
* `netbird-ui` should work with `services.netbird.enable = true;` out of the box

nazarewk commented:
Actually `netbird-ui` seems to ignore `netbird` wrapper configuration and rolls back the port to `51820` no matter what I set up. Will need to resolve it tomorrow.

CertainLach commented:
This polkit rule is required when running netbird daemon as unprivileged user with systemd-resolved as dns:
```js
      polkit.addRule(function(action, subject) {
        if((action.id == "org.freedesktop.resolve1.set-dns-servers" ||
            action.id == "org.freedesktop.resolve1.set-domains") &&
           subject.user == "netbird") {
          return polkit.Result.YES;
        }
      })
```
Without this rule, netbird is failing to reconfigure DNS records, and DNS doesn't work.

I also have patched netbird bpf programs in my setup, because netbird uses bpf_trace_printk, which is a privileged system call (Because it might slow down the system), without this patch, bpf program is failing to load, and netbird fallbacks to userspace wireguard implementation.

nazarewk commented:
> This polkit rule is required when running netbird daemon as unprivileged user with systemd-resolved as dns:

great catch, i completely missed it when copying my configs https://github.com/nazarewk-iac/nix-configs/commit/eaba33c906d26ef01a2b2cd5686ec8c2f8aeaca2#diff-2e9b46dc688a39f84859f3f7e750ffa4290c4eaf8a07f5e531cbb9afb1ab040bL182

nazarewk commented:
moving back to Draft until I resolve following:
- fix polkit rules
- create `netbird-ui` wrappers

nazarewk commented:
Changes introduced:
- moved a lot of documentation from manual into options `description` and made it more comprehensive
- added `netbird-ui` wrapper: both binary and a desktop entry
- tested `netbird-ui`
- pinned wrapper to only one kind of configuration due to https://github.com/netbirdio/netbird/issues/1581

nazarewk commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,480 @@
-              NB_INTERFACE_NAME = name;
-              NB_DAEMON_ADDR = "unix:///var/run/${config.stateDir}/sock";
+            config.environment = {
+              NB_CONFIG = "/etc/${config.name}/config.json";
```
this is a good point

nazarewk commented:
As a workaround to lack of some configurations outside `netbird up` I have resorted to adding `/etc/netbird-<name>/config.d/*.json` handling in `preStart` and fixed the segfault during incomplete config startup  with https://github.com/netbirdio/netbird/pull/1586

misuzu commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,496 @@
-            { environment, stateDir, ... }:
-            nameValuePair "netbird-${name}" {
-              description = "A WireGuard-based mesh network that connects your devices into a single private network";
+      environment.etc = toClientAttrs (client: nameValuePair "${client.name}/config.d/50-nixos.json" {
```
Why not use `builtins.toJSON client.config` expression directly in the `preStart` script?

nazarewk replied:
I want to be able to add additional configs (eg credentials) outside the NixOS configuration

misuzu replied:
It's fine to add a workaround until the issue is fixed but I don't think it's a good idea to advertise ways of configuration which are not supported by upstream

nazarewk replied:
good idea, made it explicit in `config` option description.

Tom-Hubrecht replied:
As the configuration is a json file, we can use https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/utils.nix#L160 to manage secrets in it without putting them in the nix store

misuzu commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,504 @@
+                  Additional configuration that exists before the first start and
+                  later overrides the existing values in `config.json`.
+
+                  WARNING: this is not an upstream feature, see below for details.
```
I'd emphasize that it can break at any moment

nazarewk replied:
I would see an upstream feature built on top of [`ConfigInput` structure ](https://github.com/netbirdio/netbird/pull/1586/files#diff-a669c9314247b993250fdcde6c802c8f689df6e5ce935ad890a4548c7b380a9aR39) currently used for modifying configs by CLI flags.

It would use a different format and possibly have higher priority due to being executed later in the startup process.

This NixOS option would still be useful for configurations not supported by `ConfigInput`.

misuzu replied:
The only valid way of configuring netbird is via environment variables or cli options.The `config.json` is not a config, but a state file and it's expected to be modified only by netbird itself. If we expose a footgun it should be marked as such.

misuzu commented on nixos/modules/services/networking/netbird.md:
```diff
@@ -2,55 +2,78 @@
-The `enable` option is mainly kept for backward compatibility, as defining netbird
-tunnels through the `tunnels` option is more expressive.
+This will set up a `netbird.service` listening on the port `51820` associated to the
+`wt0` interface. There will also be `netbird-wt0` binary installed in addition to `netbird`.
```
Using the config from https://docs.netbird.io/how-to/installation#linux
```nix
 { config, pkgs, ... }:
 {
   services.netbird.enable = true; # for netbird service & CLI
   environment.systemPackages = [ pkgs.netbird-ui ]; # for GUI
 }
```

And then running as described in https://docs.netbird.io/how-to/installation#running-net-bird-with-sso-login it doesn't work as expected:
```
% netbird up
netbird: command not found
```

nazarewk replied:
created an alias for `netbird` and `netbird-ui` command

misuzu commented:
The NixOS test looks broken:
```
cleanup
kill machine (pid 9)
qemu-kvm: terminating on signal 15 from pid 6 (/nix/store/3v2ch16fkl50i85n05h5ckss8pxx6836-python3-3.11.8/bin/python3.11)
(finished: cleanup, in 0.01 seconds)
Traceback (most recent call last):
  File "/nix/store/1nvy97rv9vcc9fz51yf4m6n1xn1pz43a-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
    sys.exit(main())
             ^^^^^^
  File "/nix/store/1nvy97rv9vcc9fz51yf4m6n1xn1pz43a-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/__init__.py", line 126, in main
    driver.run_tests()
  File "/nix/store/1nvy97rv9vcc9fz51yf4m6n1xn1pz43a-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/driver.py", line 163, in run_tests
    self.test_script()
  File "/nix/store/1nvy97rv9vcc9fz51yf4m6n1xn1pz43a-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/driver.py", line 155, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 2, in <module>
  File "/nix/store/1nvy97rv9vcc9fz51yf4m6n1xn1pz43a-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/machine.py", line 374, in wait_for_unit
    retry(check_active, timeout)
  File "/nix/store/1nvy97rv9vcc9fz51yf4m6n1xn1pz43a-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/machine.py", line 129, in retry
    if fn(False):
       ^^^^^^^^^
  File "/nix/store/1nvy97rv9vcc9fz51yf4m6n1xn1pz43a-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/machine.py", line 364, in check_active
    raise Exception(
Exception: unit "netbird-wt0.service" is inactive and there are no pending jobs
kill vlan (pid 7)
```

nazarewk commented:
> The NixOS test looks broken:

yeah, looks like it expects `netbird-wt0.service`, fixing it

misuzu commented:
Fails to evaluate on my machine...
```
       error: A definition for option `environment.systemPackages."[definition 27-entry 1]"' is not of type `package'. Definition values:
       - In `/nix/store/qzh54zpssd9xi1fxg9zdiqpfc892c5xa-source/nixos/modules/services/networking/netbird.nix': <function>
```

nazarewk commented:
> The NixOS test looks broken:

seems like there is no `NeedLogin` anymore in the status, fixing the test now

nazarewk commented:
just an FYI: I am fiddling with DNS servers on my workplace Netbird instance (which gets connected on-demand - always after my private instance) and seems like the DNS services of Netbird clients are completely exclusive:
- not even non-clashing DNS domains work - only the first client started serves ANY queries through systemd
- the workplace instance works fine after I disabled my private instance and restarted it

misuzu commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,516 @@
-    (mkIf (cfg.tunnels != { }) {
-      boot.extraModulePackages = optional (versionOlder kernel.kernel.version "5.6") kernel.wireguard;
+        environment.systemPackages = [
+          (pkgs.runCommand "${client.wrapper}-as-default" { } ''
```
```suggestion
          (pkgs.runCommand "${name}-as-default" { } ''
```

Otherwise we'll get weird paths like `/nix/store/x0b6rd68r5d2z075i8yl3rpmsapsr6v8--nix-store-ia8f00xq8ywm0mpncphly51fvg3hfcdp-netbird-0.26.3-wrapper-wt0-as-default/bin/netbird`

misuzu replied:
It also makes sense to add `lib.hiPrio` to make sure that wrapper gets in the PATH instead of directly added package (the `environment.systemPackages = [ pkgs.netbird-ui ];` case)

misuzu commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,516 @@
+    concatLists
+    concatStringsSep
+    escapeShellArgs
+    filter
```
`filter`, `optionals` and `removePrefix` could be removed

misuzu reviewed (commented):
LGTM otherwise

misuzu commented:
> seems like there is no `NeedLogin` anymore in the status

It should though. This is how it looks like on master:
```
Daemon status: NeedsLogin

Run UP command to log in with SSO (interactive login):

 netbird up 

If you are running a self-hosted version and no SSO provider has been configured in your Management Server,
you can use a setup-key:

 netbird up --management-url <YOUR_MANAGEMENT_URL> --setup-key <YOUR_SETUP_KEY>

More info: https://docs.netbird.io/how-to/register-machines-using-setup-keys
```

nazarewk commented:
> > seems like there is no `NeedLogin` anymore in the status
> 
> It should though. This is how it looks like on master:

Weird as it is NixOS tests have different output (probably different stage of discovery on the first run since some time), after adding `assert False, output` it's:
```
...
  File "<string>", line 5, in <module>
AssertionError: Daemon version: 0.26.3
CLI version: 0.26.3
Management: Disconnected
Signal: Disconnected
Relays: 0/0 Available
Nameservers: 0/0 Available
FQDN: 
NetBird IP: N/A
Interface type: N/A
Quantum resistance: false
Routes: -
Peers count: 0/0 Connected

kill vlan (pid 7)
```

nazarewk commented on pkgs/tools/networking/netbird/default.nix:
```diff
@@ -58,6 +59,14 @@ buildGoModule rec {
     WebKit
   ];
 
+  patches = [
```
yeah, I just never got to the point where the PR is about to be merged and juggling 1 commit is a lot easier than 2

nazarewk committed:
netbird: 0.35.2 -> 0.36.3

misuzu commented:
> probably caused by NixOS tests being run offline?

Tested on https://github.com/NixOS/nixpkgs/commit/d8fe5e6c92d0d190646fb9f1056741a229980089:

```
% git diff                            
diff --git a/nixos/tests/netbird.nix b/nixos/tests/netbird.nix
index 7342e8d04a39..7ef32db410f9 100644
--- a/nixos/tests/netbird.nix
+++ b/nixos/tests/netbird.nix
@@ -16,6 +16,7 @@ import ./make-test-python.nix ({ pkgs, lib, ... }:
     start_all()
     node.wait_for_unit("netbird-wt0.service")
     node.wait_for_file("/var/run/netbird/sock")
-    node.succeed("netbird status | grep -q 'Daemon status: NeedsLogin'")
+    output = node.succeed("netbird status")
+    assert False, output
   '';
 })
% nix-build . -A nixosTests.netbird
<................................................>
node # [    9.672276] netbird[801]: 2024-04-02T12:23:04Z INFO client/cmd/service_controller.go:64: started daemon server: /var/run/netbird/sock
(finished: waiting for unit netbird-wt0.service, in 10.77 seconds)
node: waiting for file '/var/run/netbird/sock'
(finished: waiting for file '/var/run/netbird/sock', in 0.03 seconds)
node: must succeed: netbird status
(finished: must succeed: netbird status, in 0.08 seconds)
cleanup
kill machine (pid 9)
qemu-kvm: terminating on signal 15 from pid 6 (/nix/store/7wz6hm9i8wljz0hgwz1wqmn2zlbgavrq-python3-3.11.8/bin/python3.11)
(finished: cleanup, in 0.01 seconds)
Traceback (most recent call last):
  File "/nix/store/k6yj2za3m4zg5mmviq23h0ir7y604fp6-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
    sys.exit(main())
             ^^^^^^
  File "/nix/store/k6yj2za3m4zg5mmviq23h0ir7y604fp6-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/__init__.py", line 126, in main
    driver.run_tests()
  File "/nix/store/k6yj2za3m4zg5mmviq23h0ir7y604fp6-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/driver.py", line 163, in run_tests
    self.test_script()
  File "/nix/store/k6yj2za3m4zg5mmviq23h0ir7y604fp6-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/driver.py", line 155, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 5, in <module>
AssertionError: Daemon status: NeedsLogin

Run UP command to log in with SSO (interactive login):

 netbird up 

If you are running a self-hosted version and no SSO provider has been configured in your Management Server,
you can use a setup-key:

 netbird up --management-url <YOUR_MANAGEMENT_URL> --setup-key <YOUR_SETUP_KEY>

More info: https://docs.netbird.io/how-to/register-machines-using-setup-keys


kill vlan (pid 7)
```

nazarewk commented:
> > probably caused by NixOS tests being run offline?
> 
> Tested on [d8fe5e6](https://github.com/NixOS/nixpkgs/commit/d8fe5e6c92d0d190646fb9f1056741a229980089):

Then it must be caused by my patch making daemon think it was already configured.

misuzu commented on pkgs/tools/networking/netbird/default.nix:
```diff
@@ -58,6 +59,14 @@ buildGoModule rec {
+  patches = [
+    # Accept incomplete configs
+    (fetchpatch2 {
+      url = "https://github.com/netbirdio/netbird/pull/1586/commits/1a7e2051db312ed236e468ea0d2371bde24a393a.patch";
```
The patch needs a rebase:
```
error: builder for '/nix/store/ckqbdyj4r45x3fk9346q0aa8jx1ld01c-netbird-0.27.1.drv' failed with exit code 1;
       last 10 log lines:
       > unpacking source archive /nix/store/0rm9vn9vakgn2a8kza6bb6nrs7mbpx49-source
       > source root is source
       > Running phase: patchPhase
       > applying patch /nix/store/fczbw1dbsx2gh247p6h56pxdqm4gi33d-1a7e2051db312ed236e468ea0d2371bde24a393a.patch
       > patching file client/internal/config.go
       > Hunk #2 succeeded at 49 with fuzz 1 (offset 2 lines).
       > Hunk #3 succeeded at 68 (offset 3 lines).
       > Hunk #4 succeeded at 108 (offset 3 lines).
       > Hunk #5 FAILED at 165.
       > 1 out of 5 hunks FAILED -- saving rejects to file client/internal/config.go.rej
       For full logs, run 'nix log /nix/store/ckqbdyj4r45x3fk9346q0aa8jx1ld01c-netbird-0.27.1.drv'.
```

oddlama commented:
@nazarewk Can you rebase this to address the conflicts? This is a great addition and I would love to see this merged!

oddlama commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,513 @@
     ;
 
   inherit (lib.types)
+    anything
```
This seems to be unused

oddlama commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,513 @@
+          - `netbird-<name>` script,
+          - `Netbird - netbird-<name>` graphical interface when appropriate (see `ui.enable`),
+        2. run as a `netbird-<name>.service`,
+        3. listen for incoming remote connections on the port `51830` (`openFirewall` by default),
```
This is a typo? Looks like the default is port `51820`

oddlama commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,513 @@
+                '';
+              };
+
+              interface = mkOption {
```
Interfaces names have a maximum length of 15 characters on linux. Since the default `netbird-${name}` is already quite long and only leaves 7 characters for the $name, can we add an assertion that this limit isn't exceeded? Maybe it would also be better to choose a shorter default prefix.

nazarewk replied:
wasn't aware of that 15 character limit, thanks for pointing it out

oddlama commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,513 @@
-                    NB_WIREGUARD_PORT = builtins.toString port;
-                    NB_INTERFACE_NAME = name;
-                    NB_DAMEON_ADDR = "/var/run/''${stateDir}"
+                    NB_CONFIG = "/var/lib/''${config.name}/config.json";
```
I think freely letting the user choose `name` here is problematic, because it is directly used as a directory name in `/var/lib` or `/run`, which isn't something a user would expect.

If I choose a name similar to another service on the system, their state directories will clash which can cause a lot of issues that we should avoid IMO by always forcing the `netbird-` prefix everywhere $name is used right now, instead of just setting the default to `netbird-${<attrname>}`.

oddlama replied:
Maybe introducing a `stateDir` option that defaults to `netbird-${name}` would suffice.

nazarewk replied:
actually forcing a prefix sounds like a good idea

oddlama commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,513 @@
-              path = with pkgs; [ openresolv ];
+        documentation = [ "https://netbird.io/docs/" ];
 
-              inherit environment;
```
This should be added again, currently the environment is not passed to the service at all. So most of the default settings will have no effect. (the exports from the preStart scripts don't automatically pass along to the main service)

oddlama replied:
Never mind, ignore this. I overlooked the make-wrapper set-default that you used.

oddlama commented:
This new module unfortunately causes netbird to segfault on start, except when you use the old enable option. This seems to be because `/etc/netbird/config.json` contains correct addresses for management url and admin url, while configurations with other directories will have them default to `null` which causes a segfault because url structs are not expected to be null.

Can you add tests to ensure this doesn't regress in the future?

Example configuration that fails:
```nix
  services.netbird.clients.test = {
    port = 51820;
    name = "netbird-test";
    interface = "wt-test";
    environment = {
      NB_MANAGEMENT_URL = "https://mynetbird.example.com";
    };
  };
```

```nix
2024-05-18T00:43:39+02:00 INFO client/cmd/service_controller.go:24: starting Netbird service
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0xf60ebd]

goroutine 13 [running]:
net/url.(*URL).Hostname(...)
	net/url/url.go:1151
github.com/netbirdio/netbird/client/internal.UpdateOldManagementURL({0x174ac80, 0xc0006a4460}, 0xc0006dab00, {0xc0000440ca, 0x21})
	github.com/netbirdio/netbird/client/internal/config.go:408 +0xfd
github.com/netbirdio/netbird/client/server.(*Server).Start(0xc00056da40)
	github.com/netbirdio/netbird/client/server/server.go:131 +0x4e5
github.com/netbirdio/netbird/client/cmd.(*program).Start.func1()
	github.com/netbirdio/netbird/client/cmd/service_controller.go:59 +0x4e6
created by github.com/netbirdio/netbird/client/cmd.(*program).Start in goroutine 1
	github.com/netbirdio/netbird/client/cmd/service_controller.go:47 +0x315
```

oddlama commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,513 @@
-    tunnels = mkOption {
+    ui.enable = mkOption {
+      type = bool;
+      default = config.services.xserver.displayManager.sessionPackages != [ ];
```
Is there any rationale for using `config.services.xserver.displayManager.sessionPackages != [ ]` over `config.services.xserver.enable`?

If yes, note that `config.services.xserver.displayManager.sessionPackages` has been renamed to `config.services.displayManager.sessionPackages`

nazarewk replied:
yeah, I just realized the source of my month-long errors about using renamed option.
checking for `sessionPackages` makes more sense because those can also contain Wayland sessions, not just X sessions.

oddlama replied:
On my configuration the sessionPackages are empty because I use startx and no display manager. So maybe we can use ` || config.services.xserver.enable` as an additional clause?

oddlama commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,513 @@
+                        makeWrapper ${lib.getExe cfg.package} "$out/bin/netbird-${name}" \
+                          ${escapeShellArgs makeWrapperArgs}
+                      ''
+                      (optionalString cfg.ui.enable ''
```
The client specific `ui.enable` option is never used, because this is the global option

nazarewk replied:
yeah, you either build UI wrapper for all clients or none at all

PatrickDaG replied:
Should the `client.ui.enable` option then be removed as it doesn't do anything?

oddlama commented:
One more thing: When enabling `ui.enable`, only the `netbird-<name>` is added to the system path. The ui wrapper `netbird-ui-<name>` will not be easily accessible, except for the old `netbird.enable` where you add an explicit "as-default" wrapper.

nazarewk commented:
> This new module unfortunately causes netbird to segfault on start, except when you use the old enable option. This seems to be because `/etc/netbird/config.json` contains correct addresses for management url and admin url, while configurations with other directories will have them default to `null` which causes a segfault because url structs are not expected to be null.
> ...
> **EDIT:** The culprit actually seems to be related to the prestart script. If the config.json file does not exist yet, netbird will usually populate it using the correct values. But if it doesn't exist before running the pre-start script, it will be created containing only a subset of the required values, causing netbird to assume they should be unset or something.

might be caused by merging/lack of https://github.com/netbirdio/netbird/pull/1586 , I'll simply rebase today and try to address your comments tomorrow.

nazarewk commented:
> When enabling `ui.enable`, only the `netbird-<name>` is added to the system path. The ui wrapper `netbird-ui-<name>` will not be easily accessible, except for the old `netbird.enable` where you add an explicit "as-default" wrapper.

Works fine for me:
```
> which netbird-ui-priv netbird-ui-sc
/run/current-system/sw/bin/netbird-ui-priv
/run/current-system/sw/bin/netbird-ui-sc
```

oddlama commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,171 +1,518 @@
+              name = mkOption {
+                type = str;
+                default = name;
+                apply = value: "netbird-${value}";
```
I think it would be better style to force the prefix in the relevant locations, otherwise `/var/lib/${cfg.name}` will always look problematic since it implies any name is possible.

oddlama reviewed (commented):
I think you'll have to adjust the release notes to apply to 24.11 now that 24.05 is out the door. But the important details look fine to me, so this is good to go from my side. We can iron out any possible details separately if necessary.

oddlama commented:
There are still changes to `nixos/doc/manual/release-notes/rl-2405.section.md`

nazarewk commented:
> There are still changes to `nixos/doc/manual/release-notes/rl-2405.section.md`

Yeah, the nixos manual command refused to pass, assumed it was expecting me to change it?

oddlama commented:
> Yeah, the nixos manual command refused to pass, assumed it was expecting me to change it?

I guess it complains because you cannot edit old release notes

nazarewk commented:
> > Yeah, the nixos manual command refused to pass, assumed it was expecting me to change it?
> 
> I guess it complains because you cannot edit old release notes

it refused to pass without changes https://github.com/NixOS/nixpkgs/actions/runs/9399057023/job/25885722356

PatrickDaG commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,168 +1,549 @@
+      ));
 
-              documentation = [ "https://netbird.io/docs/" ];
+      environment.etc = toClientAttrs (client: nameValuePair "netbird-${client.name}/config.d/50-nixos.json" {
```
Do you plan to add more override files, or why do you first write the file in `/etc` and later iterate over the one file? I think it would be simpler to skip this and directly access the file from the nix-store in the derivation, skipping one indirection.

nazarewk replied:
I'm using those on some of my machines and while testing stuff.

PatrickDaG commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -1,168 +1,549 @@
 
-              documentation = [ "https://netbird.io/docs/" ];
+      environment.etc = toClientAttrs (client: nameValuePair "netbird-${client.name}/config.d/50-nixos.json" {
+        text = builtins.toJSON client.config;
```
Probs would be better to use ` formatType.generate "config.json" cfg.config;` since you're using the json format type already

nazarewk replied:
don't think that's necessary, I think the generation uses `jq` with a separate derivation underneath.

PatrickDaG reviewed (approved):
Module looks great. Hope it gets merged soon, been long enough.
I would like to test this more myself, but I'm working on a rewrite of the [netbird-server](https://github.com/NixOS/nixpkgs/pull/354032) modules right now and while I don't think they actually conflict, the merges do.

nazarewk commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -711,7 +711,7 @@ Use `services.pipewire.extraConfig` or `services.pipewire.configPackages` for Pi
   512+ bits instead of 128+ bits.
 
-- `services.netbird` now allows running multiple tunnels in parallel through [`services.netbird.tunnels`](#opt-services.netbird.tunnels).
+- `services.netbird` now allows running multiple tunnels in parallel through [`services.netbird.tunnels`](#opt-services.netbird.clients).
```
to be honest, this PR is hanging for so many months I lost hope it will ever get merged and I stopped trying to chase people able to merge it.

nazarewk replied:
I'll make the "proper" realease notes when some of the maintainers able to give it review and merge pops up.
The code itself was written around 23.11 and already missed 2 releases.

oddlama replied:
I know how frustrating a long running PR can be. The unfortunate thing is that any committer who comes along will see that this is missing proper release notes, and then can't merge it. So this will only delay it further.

I fully understand if you say that this requires too much work to keep up in the long term. Let me know if you need support with anything, otherwise I'd also offer pick this PR up if you don't want to keep working on it.

fricklerhandwerk replied:
@nazarewk sorry for the long delay on this PR. Unfortunately maintainers' resources are very tightly limited and unpredictable as almost everyone here is a volunteer doing this on their own time, so all of that is best effort -- sometimes this results in months and months of inactivity.

The correct way to address the changed option type is using `mkRenamedOptionModule` in `netbird.nix`:

```nix
  imports = [
    (lib.mkRenamedOptionModule [ "services" "netbird" "tunnels" ] [ "services" "netbird" "clients" ])
  ];
```

Old release notes should never be changed.

nazarewk replied:
`(mkRenamedOptionModule [ "services" "netbird" "tunnels" ] [ "services" "netbird" "clients" ])` does not seem to work:
- failing check: https://github.com/NixOS/nixpkgs/actions/runs/12927554939/job/36052926706?pr=287236#step:5:241
- the code: https://github.com/NixOS/nixpkgs/blob/c65c0c24c0b1fe683cea50f1c32893304cf0b1f7/nixos/modules/services/networking/netbird.nix#L75-L77

fricklerhandwerk replied:
Ah yes, as I've found there are a lot of variants of that helper in `lib/modules.nix`. We need one that renders the old option to the manual, so the original anchor is created. It seems that there's none which both renders the alias to the manual *and* throws an evaluation warning, so eventually we should add that. To get this PR in, please use `mkAliasOptionModule` -- this will allow users to use any of those names, and we'll hopefully fix up the warning/rendering behavior until the next stable release.

Btw not changing release notes was not monitored and couldn't be enforced without the redirects mechanism that now grinds us down (for the right reasons).

PatrickDaG commented on nixos/modules/services/networking/netbird.nix:
```diff
@@ -7,162 +7,596 @@
-                  Directory storing the netbird configuration.
+                  Start the service with the system.
+
+                  As of 2024-02-13 it is not possible to start a Netbird client daemon without immediately
```
There is an `disable-auto-connect` flag but it seems to only apply to `netbird up`, the link however seems to have run stale as there is no mention of such a feature being planned as far as I can see.

nazarewk replied:
gotta look into it, I'm using some form of this in my configs successfully

nazarewk committed:
nixos/netbird: harden and extend options

nazarewk committed:
nixos/netbird: update docs & release notes

fricklerhandwerk commented:
I'd merge but NixOS tests seem to be broken or hanging? @Mic92 how to unblock this?

oddlama commented:
(And this still modifies old release notes)

Mic92 commented:
> I'd merge but NixOS tests seem to be broken or hanging? @Mic92 how to unblock this?

Fixing the test? I am not using netbird and have currently other things on my list. Sorry.

nazarewk commented:
> > I'd merge but NixOS tests seem to be broken or hanging? @Mic92 how to unblock this?
> 
> Fixing the test? I am not using netbird and have currently other things on my list. Sorry.

I'll adress remaining things this week. Otherwise I didn't touch it apart from rebasing for months already.

nazarewk commented:
I have fixed the tests (I did some incompatible changes to the module on the way), but even though I've added option rename, the manual still doesn't build:
https://github.com/NixOS/nixpkgs/blob/c65c0c24c0b1fe683cea50f1c32893304cf0b1f7/nixos/modules/services/networking/netbird.nix#L75-L77

nazarewk committed:
netbird-ui: fix meta.mainProgram

nazarewk commented:
> > I have fixed the tests (I did some incompatible changes to the module on the way), but even though I've added option rename, the manual still doesn't build:
> > https://github.com/NixOS/nixpkgs/blob/c65c0c24c0b1fe683cea50f1c32893304cf0b1f7/nixos/modules/services/networking/netbird.nix#L75-L77
> 
> seems like there is some (actually quite a lot, just not AS relevant as this one) precedent to removing & editing old release notes: [e3812e1#diff-9538c800780031db3dfa7746f5a36fbbc895c60fdc896385bedd36940927427dL79-L81](https://github.com/NixOS/nixpkgs/commit/e3812e187592c866349630cc097c786b353130c7#diff-9538c800780031db3dfa7746f5a36fbbc895c60fdc896385bedd36940927427dL79-L81)

As per https://github.com/NixOS/nixpkgs/pull/287236#discussion_r1930212710 , I am temporarily using `mkAliasOptionModule` until there is a better implementation to build release notes without errors.

fricklerhandwerk commented:
Thanks a great deal for sticking through it, and sorry for the long delay.
